subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/ToolingModelBuilderResultInternal.java

```diff
@@ -16,30 +16,45 @@
 
 package org.gradle.tooling.provider.model.internal;
 
+import com.google.common.collect.ImmutableList;
 import org.gradle.internal.problems.failure.Failure;
 import org.gradle.tooling.provider.model.ToolingModelBuilder;
 import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
 
-import java.util.Collections;
 import java.util.List;
 
 /**
  * Internal type for a model builder result that contains additional data beyond the model itself if needed.
  * <p>
  * {@link ToolingModelBuilder} can optionally return a {@link ToolingModelBuilderResultInternal} instead of a raw model.
  * {@link ToolingModelBuilderResultInternal} is then unpacked before the model is returned to the client.
+ * <p>
+ * <strong>This class is intentionally not {@code Serializable}</strong>. It must not travel over the Tooling API serialization boundary.
  */
 @NullMarked
 public class ToolingModelBuilderResultInternal {
 
-    private final Object model;
+    private final @Nullable Object model;
     private final List<Failure> failures;
 
-    private ToolingModelBuilderResultInternal(Object model, List<Failure> failures) {
+    private ToolingModelBuilderResultInternal(@Nullable Object model, List<Failure> failures) {
         this.model = model;
         this.failures = failures;
     }
 
+    public RuntimeException throwOriginal() {
```

Comment:
💭   We could also change this method to `throwFailureIfPresent()`

And logic would be:
```
if (failures.isEmpty()) {
   return;
}
throw failure logic...
```

That way you wouldn't need to do a check `if (!result.failures.isEmpty())` in places where you want to throw if there is a failure.